akka-stream-tests/src/test/scala/akka/stream/impl/GraphStageLogicSpec.scala

```diff
@@ -155,7 +164,13 @@ class GraphStageLogicSpec extends StreamSpec with GraphInterpreterSpecKit {
 
       Source.empty.via(g).runWith(TestSink.probe)
         .request(9)
-        .expectNextN(1 to 8)
+        .expectNext(1)
+        //emitting with callback gives nondeterminism whether 2 or 3 will be pushed first
+        .expectNextUnordered(2, 3)
+        .expectNext(4)
+        .expectNext(5)
+        .expectNextUnordered(6, 7)
+        .expectNext(8)
```

Comment:
Not quite, it's interesting. 
In fact, for emit1234 the actual order was always 1324 (but due to corruption test was passing with 1234) because emission is done in preStart(). 
As I described, sequence of actions was as following:
- after emit(out, 1, () ⇒ emit(out, 3)) OuputHandler was set to new EmittingSingle(out, 1, getNonEmittingHandler(out)
- then emit(out, 2, () ⇒ emit(out, 4)) was called and new EmittingSingle(out, 2, getNonEmittingHandler(out) was added to previous Emitting to the tail
- then () ⇒ emit(out, 3) was invoked and new EmittingSingle(out, 3, getNonEmittingHandler(out), andThen) prepended the resulting Emitting[] handler by the followUp call, causing 3 to be pushed before 2

For emit5678 it's more complicated, the actual order of elements is really nondeterministic 5768 or 5678 either. 
But tests always worked fine because of following:
- for 5768 it was similar to 1324
- for 5678 it worked fine because this happens when IN had been pulled before emit(5) was invoked and emit uses, another branch where it immediately pushes 5, then immediately invokes () => emit(6), and then it calls emit(7), then when 7 was emitted and () => emit(8) was invoked there were no followUps and 8 was emitted in right order.